goflags.go

```diff
@@ -342,35 +374,102 @@ func (stringSlice *StringSlice) createStringArrayDefaultValue() string {
 }
 
 func (flagSet *FlagSet) usageFunc() {
-	hashes := make(map[string]struct{})
-
 	cliOutput := flag.CommandLine.Output()
 	fmt.Fprintf(cliOutput, "%s\n\n", flagSet.description)
 	fmt.Fprintf(cliOutput, "Usage:\n  %s [flags]\n\n", os.Args[0])
 	fmt.Fprintf(cliOutput, "Flags:\n")
 
 	writer := tabwriter.NewWriter(cliOutput, 0, 0, 1, ' ', 0)
 
-	flagSet.flagKeys.forEach(func(key string, data *flagData) {
+	if len(flagSet.groups) > 0 {
+		flagSet.usageFuncForGroups(cliOutput, writer)
+	} else {
+		flagSet.usageFuncInternal(cliOutput, writer)
+	}
+}
+
+// usageFuncInternal prints usage for command line flags
+func (flagSet *FlagSet) usageFuncInternal(cliOutput io.Writer, writer *tabwriter.Writer) {
+	uniqueIterator := newUniqueFlagIterator()
+
+	flagSet.flagKeys.forEach(func(key string, data *FlagData) {
 		currentFlag := flag.CommandLine.Lookup(key)
 
-		dataHash := data.Hash()
-		if _, ok := hashes[dataHash]; ok {
-			return // Don't print the value if printed previously
+		if !uniqueIterator.validate(data) {
+			return
 		}
-		hashes[dataHash] = struct{}{}
-
 		result := createUsageString(data, currentFlag)
 		fmt.Fprint(writer, result, "\n")
 	})
 	writer.Flush()
 }
 
+// usageFuncForGroups prints usage for command line flags with grouping enabled
+func (flagSet *FlagSet) usageFuncForGroups(cliOutput io.Writer, writer *tabwriter.Writer) {
+	uniqueIterator := newUniqueFlagIterator()
+
+	var otherOptions []string
+	for _, group := range flagSet.groups {
+		fmt.Fprintf(cliOutput, "%s:\n", normalizeGroupDescription(group.description))
+
+		flagSet.flagKeys.forEach(func(key string, data *FlagData) {
+			if data.group == "" {
+				if !uniqueIterator.validate(data) {
+					return
+				}
+				currentFlag := flag.CommandLine.Lookup(key)
```

Comment:
You could extract the `currentFlag` before the if statement, and reuse it in line 432, instead of querying the same value twice.